src/collector.py

```diff
@@ -0,0 +1,456 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""OpenSearch Dashboards Collector."""
+
+import json
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from typing import Generator, Optional
+
+import requests
+from prometheus_client.core import GaugeMetricFamily, Metric
+from prometheus_client.registry import Collector
+from requests.auth import HTTPBasicAuth
+from requests.exceptions import HTTPError, RequestException, Timeout
+
+METRICS_PREFIX = "opensearch_dashboards_"
+API_STATUS_ENDPOINT = "/api/status"
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Config:
+    """Wrap CLI arguments and environment variables necessary to get metrics from the API."""
+
+    url: str
+    user: str
+    password: str
+
+
+class Heap(Enum):
+    """Possible heap types for the metrics"""
+
+    USED = "used_in_bytes"
+    TOTAL = "total_in_bytes"
+    SIZE = "size_limit"
+
+
+class Load(Enum):
+    """Possible heap types for the metrics"""
+
+    ONE_M = "1m"
+    FIVE_M = "5m"
+    FIFTEEN_M = "15m"
+
+
+class Memory(Enum):
+    """Possible memory types for the metrics"""
+
+    USED = "used_in_bytes"
+    TOTAL = "total_in_bytes"
+    FREE = "free_in_bytes"
+
+
+class Response(Enum):
+    """Possible response times types for the metrics"""
+
+    MAX = "max_in_millis"
+    AVG = "avg_in_millis"
+
+
+class RequestsCount(Enum):
+    """Possible requests count types for the metrics"""
+
+    DISCONNECTS = "disconnects"
+    TOTAL = "total"
+
+
+class DashBoardsCollector(Collector):
+    """OpenSearch Dashboards Collector"""
+
+    def __init__(self, config: Config) -> None:
+        super().__init__()
+        self.config = config
+
+    def collect(self) -> Generator[Metric, None, None]:
+        """Collect all the metrics from /api/status to be exported.
+
+        Yields:
+            Generator[Metric]: the internal status metrics of status
+        """
+        api_up = GaugeMetricFamily(
+            name=f"{METRICS_PREFIX}up",
+            documentation="Whether the data source is reachable (1 for up, 0 for down)",
+        )
+        try:
+            if api_metrics := collect_api_status(self.config):
+                api_up.add_metric([], 1)
+                for metric in self.metrics(api_metrics):
+                    description, value = metric
+                    if value:
+                        yield value
+                    else:
+                        logger.error("It was not possible to get the metric: %s", description)
+            else:
+                api_up.add_metric([], 0)
+
+        except Exception as e:  # pylint: disable=broad-except
```

Comment:
**todo:** I don't think you should broad-except here. Let it fail if you don't handle the correct exception, and you can handle that after you find out what it was. Better that than just catching everything.

Reply:
I'll double check, but my concern here is if I don't catch an exception the exporter itself might crush. I think it's easier to create a Prometheus alert to check that opensearch_dashboards_up is 0 (something really bad happened) than checking if the metrics are absent because we can have multiple opensearch units and some might work and some might not.

Reply:
I gave a second thought and if exporter crashes, prometheus would start return `0` for the up metric and a alert would trigger because of that. This rule was introduced here: https://github.com/canonical/opensearch-dashboards-operator/pull/140

With that said, I think I agree that we can let it crash, instead of having this broad catch.

Thanks

Reply:
Yeah, I agree a general catch will lead to the exporter _not_ crashing when it would be the expected thing. I would limit it to the typical requests exceptions on timeouts, unreachable endpoints...